overlord/devicestate/devicestate_cloudinit_test.go

```diff
@@ -0,0 +1,951 @@
+}
+
+func (s *cloudInitSuite) TestCloudInitAlreadyDisabledDoesNothing(c *C) {
+	// the absence of a zzzz_snapd.cfg file will indicate that it has not been
+	// restricted yet and thus it should then check to see if it was manually
+	// disabled
+
+	// write a cloud-init disabled file
+	disableFile := filepath.Join(dirs.GlobalRootDir, "/etc/cloud/cloud-init.disabled")
+	err := os.MkdirAll(filepath.Dir(disableFile), 0755)
+	c.Assert(err, IsNil)
+	err = ioutil.WriteFile(disableFile, nil, 0644)
+	c.Assert(err, IsNil)
+
+	// mock cloud-init command, but make it always fail, it shouldn't be called
+	// as cloud-init.disabled should tell sysconfig to never consult cloud-init
+	// directly
+	cmd := testutil.MockCommand(c, "cloud-init", `
+echo "unexpected call to cloud-init with args $*"
+exit 1`)
+	defer cmd.Restore()
+
+	r := devicestate.MockRestrictCloudInit(func(sysconfig.CloudInitState, *sysconfig.CloudInitRestrictOptions) (sysconfig.CloudInitRestrictionResult, error) {
+		c.Error("EnsureCloudInitRestricted should not have restricted cloud-init when already disabled")
+		return sysconfig.CloudInitRestrictionResult{}, fmt.Errorf("broken")
+	})
+	defer r()
+
+	err = devicestate.EnsureCloudInitRestricted(s.mgr)
+	c.Assert(err, IsNil)
+
+	c.Assert(s.mockLogger.String(), Equals, "")
+
+	c.Assert(cmd.Calls(), HasLen, 0)
+}
+
+func (s *cloudInitSuite) TestCloudInitUntriggeredDisables(c *C) {
+	// the absence of a zzzz_snapd.cfg file will indicate that it has not been
+	// restricted yet and thus it should then check to see if it was manually
+	// disabled
+
+	// the absence of a cloud-init.disabled file indicates that cloud-init is
+	// "untriggered", i.e. not active/running but could still be triggered
+
+	cmd := testutil.MockCommand(c, "cloud-init", `
+if [ "$1" = "status" ]; then
+	echo "status: disabled"
+else
+	echo "unexpected args $*"
+	exit 1
+fi`)
+	defer cmd.Restore()
+
+	restrictCalls := 0
+
+	r := devicestate.MockRestrictCloudInit(func(state sysconfig.CloudInitState, opts *sysconfig.CloudInitRestrictOptions) (sysconfig.CloudInitRestrictionResult, error) {
+		restrictCalls++
+		c.Assert(state, Equals, sysconfig.CloudInitUntriggered)
+		c.Assert(opts, Not(IsNil))
+		c.Assert(opts.ForceDisable, Equals, false)
+		// we would have disabled it
+		return sysconfig.CloudInitRestrictionResult{Action: "disable"}, nil
+	})
+	defer r()
+
+	err := devicestate.EnsureCloudInitRestricted(s.mgr)
+	c.Assert(err, IsNil)
+
+	c.Assert(cmd.Calls(), DeepEquals, [][]string{
+		{"cloud-init", "status"},
+	})
+
+	// a message about cloud-init done and being restricted
+	c.Assert(strings.TrimSpace(s.mockLogger.String()), Matches, `.*System initialized, cloud-init reported to be in disabled state, disabled permanently.*`)
+
+	c.Assert(restrictCalls, Equals, 1)
+}
+
+func (s *cloudInitSuite) TestCloudInitDoneRestricts(c *C) {
+	// mock /etc/cloud on writable
+	err := os.MkdirAll(filepath.Join(dirs.GlobalRootDir, "etc", "cloud"), 0755)
+	c.Assert(err, IsNil)
```

Comment:
also perhaps this could be moved to the SetupTest as well, since there's no test that requires this to not be here or otherwise relies on the absence / presence of this directory specifically, always the files in this directory